src/components/ShipFit/Slot.tsx

```diff
@@ -148,6 +149,54 @@ export const Slot = (props: { type: string; index: number; fittable: boolean; ma
     [shipSnapshot, esiItem],
   );
 
+  const onDragStart = React.useCallback(
+    (e: React.DragEvent<HTMLDivElement>) => {
+      if (!esiItem) return;
+
+      e.dataTransfer.setData("application/type_id", esiItem.type_id.toString());
+      e.dataTransfer.setData("application/slot_id", esiFlag.toString());
+      e.dataTransfer.setData("application/slot_type", esiFlagType);
+    },
+    [esiItem, esiFlag, esiFlagType],
+  );
+
+  const onDragOver = React.useCallback((e: React.DragEvent<HTMLDivElement>) => {
+    e.preventDefault();
+  }, []);
+
+  const onDragEnd = React.useCallback(
+    (e: React.DragEvent<HTMLDivElement>) => {
+      e.preventDefault();
+
+      const parseNumber = (maybeNumber: string): number | undefined => {
+        const num = parseInt(maybeNumber);
+        return Number.isInteger(num) ? num : undefined;
+      };
+
+      const draggedTypeId: number | undefined = parseNumber(e.dataTransfer.getData("application/type_id"));
+      const draggedSlotId: number | undefined = parseNumber(e.dataTransfer.getData("application/slot_id"));
+      const draggedSlotType: string = e.dataTransfer.getData("application/slot_type");
+
+      if (!draggedTypeId) {
+        return;
+      }
+
+      // TODO: Move validation into ShipSnapshotProvider
```

Comment:
Soon (tm) I am going to rewrite the providers, and there will be one handling all these things. But for now, I would just remove this comment, and we will clean it up once I am done rewriting the providers :)

I do appreciate that you also see that this shouldn't belong here; but the ShipSnapshotProvider is a bit of a mess currently :D So it is what it is ;)